tests/testsuite/lto.rs

```diff
@@ -22,5 +26,150 @@ fn with_deps() {
+        .with_stderr_contains("[..]`rustc[..]--crate-name test[..]-C lto[..]`")
+        .run();
+}
+
+#[cargo_test]
+fn complicated() {
+    if !cargo_test_support::is_nightly() {
+        return;
+    }
+
+    Package::new("dep-shared", "0.0.1").publish();
+    Package::new("dep-normal2", "0.0.1").publish();
+    Package::new("dep-normal", "0.0.1")
+        .dep("dep-shared", "*")
+        .dep("dep-normal2", "*")
+        .publish();
+    Package::new("dep-build2", "0.0.1").publish();
+    Package::new("dep-build", "0.0.1")
+        .dep("dep-shared", "*")
+        .dep("dep-build2", "*")
+        .publish();
+    Package::new("dep-proc-macro2", "0.0.1").publish();
+    Package::new("dep-proc-macro", "0.0.1")
+        .proc_macro(true)
+        .dep("dep-shared", "*")
+        .dep("dep-proc-macro2", "*")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "test"
+                version = "0.0.0"
+
+                [lib]
+                crate-type = ['cdylib', 'staticlib']
+
+                [dependencies]
+                dep-normal = "*"
+                dep-proc-macro = "*"
+
+                [build-dependencies]
+                dep-build = "*"
+
+                [profile.release]
+                lto = true
+            "#,
+        )
+        .file("build.rs", "fn main() {}")
+        .file("src/main.rs", "fn main() {}")
+        .file("src/lib.rs", "")
+        .build();
+    p.cargo("build -v --release")
+        // normal deps and their transitive dependencies do not need object
+        // code, so they should have linker-plugin-lto specified
+        .with_stderr_contains("[..]`rustc[..]--crate-name dep_normal2 [..]-Clinker-plugin-lto[..]`")
+        .with_stderr_contains("[..]`rustc[..]--crate-name dep_normal [..]-Clinker-plugin-lto[..]`")
+        // build dependencies and their transitive deps don't need any bitcode,
+        // so embedding should be turned off
+        .with_stderr_contains("[..]`rustc[..]--crate-name dep_build2 [..]-Cbitcode-in-rlib=no[..]`")
+        .with_stderr_contains("[..]`rustc[..]--crate-name dep_build [..]-Cbitcode-in-rlib=no[..]`")
+        .with_stderr_contains(
+            "[..]`rustc[..]--crate-name build_script_build [..]-Cbitcode-in-rlib=no[..]`",
+        )
+        // proc macro deps are the same as build deps here
+        .with_stderr_contains(
+            "[..]`rustc[..]--crate-name dep_proc_macro2 [..]-Cbitcode-in-rlib=no[..]`",
+        )
+        .with_stderr_contains(
+            "[..]`rustc[..]--crate-name dep_proc_macro [..]-Cbitcode-in-rlib=no[..]`",
+        )
+        .with_stderr_contains("[..]`rustc[..]--crate-name test [..]--crate-type bin[..]-C lto[..]`")
+        .with_stderr_contains(
+            "[..]`rustc[..]--crate-name test [..]--crate-type cdylib[..]-C lto[..]`",
+        )
+        .run();
```

Comment:
I think it would be good to have a check for the settings on dep_shared (which is no flags AFAIK). I know checking for the absence of flags is awkward. but probably worthwhile here since that's one of the important bits of what `calculate` does.